### ISSUE TITLE
Never receive didOpen or didClose

### ISSUE THREAD

**Issue by Timmmm** [Microsoft/vscode-languageserver-node#283]
Apologies if this isn't a bug and is just my misunderstanding (to be fair, the documentation isn't explicit on a lot of things). I am writing a language protocol server in Go using [this package](https://github.com/sourcegraph/go-langserver) and it works so far except I never receive `textDocument/didOpen` and `textDocument/didClose` messages.  I *do* receive `textDocument/didChange`. Here is my response to the initialize message:

		syncOptions := lsp.TextDocumentSyncOptions{
			OpenClose: true,
			Change:    lsp.TDSKIncremental,
		}

		return lsp.InitializeResult{
			Capabilities: lsp.ServerCapabilities{
				TextDocumentSync: lsp.TextDocumentSyncOptionsOrKind{
					Options: &syncOptions,
				},
				DefinitionProvider:           false,
				DocumentFormattingProvider:   false,
				DocumentSymbolProvider:       false,
				HoverProvider:                true,
				ReferencesProvider:           false,
				WorkspaceSymbolProvider:      false,
				XWorkspaceReferencesProvider: false,
				XDefinitionProvider:          false,
				XWorkspaceSymbolByProperties: false,
				SignatureHelpProvider:        &lsp.SignatureHelpOptions{TriggerCharacters: []string{" "}},
			},
		}, nil

Which is turned into JSON in the obvious way. Am I right in thinking that I should receive `textDocument/didOpen` for any document before `textDocument/didChange`?

**Comment by rcjsuen** [comment]
@Timmmm What client is your language server connected to? Visual Studio Code?

**Comment by Timmmm** [comment]
Yep.

**Comment by Timmmm** [comment]
I turned on the tracing that Go library provides and this is what it says:

```
--> request #0: initialize: {"processId":83626,"rootPath":null,"rootUri":null,"capabilities":{"workspace":{"applyEdit":true,"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":true},"executeCommand":{"dynamicRegistration":true}},"textDocument":{"synchronization":{"dynamicRegistration":true,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":true,"completionItem":{"snippetSupport":true,"commitCharactersSupport":true}},"hover":{"dynamicRegistration":true},"signatureHelp":{"dynamicRegistration":true},"definition":{"dynamicRegistration":true},"references":{"dynamicRegistration":true},"documentHighlight":{"dynamicRegistration":true},"documentSymbol":{"dynamicRegistration":true},"codeAction":{"dynamicRegistration":true},"codeLens":{"dynamicRegistration":true},"formatting":{"dynamicRegistration":true},"rangeFormatting":{"dynamicRegistration":true},"onTypeFormatting":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true},"documentLink":{"dynamicRegistration":true}}},"trace":"off"}
<-- result #0: initialize: {"capabilities":{"textDocumentSync":{"openClose":true,"change":2},"hoverProvider":true,"signatureHelpProvider":{"triggerCharacters":[" "]}}}
--> notif: initialized: {}
2017/12/03 20:30:13 Message received:  initialized
--> notif: textDocument/didChange: {"textDocument":{"uri":"file:///.../rpt.S","version":1},"contentChanges":[]}
2017/12/03 20:32:53 Message received:  textDocument/didChange
2017/12/03 20:32:53 Changed:  file:///Users/timh/workspace/poplar/ciss/test/ciss/rpt.S
--> notif: textDocument/didChange: {"textDocument":{"uri":"file:///.../rpt.S","version":1},"contentChanges":[]}
2017/12/03 20:32:53 Message received:  textDocument/didChange
2017/12/03 20:32:53 Changed:  file:///Users/timh/workspace/poplar/ciss/test/ciss/rpt.S
--> notif: textDocument/didChange: {"textDocument":{"uri":"file:///.../rpt.S","version":2},"contentChanges":[{"range":{"start":{"line":31,"character":17},"end":{"line":31,"character":17}},"rangeLength":0,"text":"\n  "}]}
--> notif: textDocument/didChange: {"textDocument":{"uri":"file:///.../rpt.S","version":2},"contentChanges":[{"range":{"start":{"line":31,"character":17},"end":{"line":31,"character":17}},"rangeLength":0,"text":"\n  "}]}
```

As you can see, no `didOpen` before the `didChange`. The `lsp-sample` does receive the `didOpen` notification, but there's no difference between my server capabilities.

**Comment by rcjsuen** [comment]
@Timmmm What if instead of using a `TextDocumentSyncOptions` object you simply used the `TextDocumentSyncKind.Incremental` numeric value.

**Comment by Timmmm** [comment]
This is really frustrating. I've added `console.log()`'s to the VSCode code and it seems to sent `initialized` twice, and `didOpen` once. My program *definitely* only receives one `initialized` and no `didOpen` so I'm guessing it is sending it somewhere else, but I can't work out how to get the tracing working.

**Comment by rcjsuen** [comment]
@Timmmm Maybe you can try reaching out to the [sourcegraph/go-langserver](https://github.com/sourcegraph/go-langserver) folks?

**Comment by Timmmm** [comment]
Ok I think what is happening is that the server is started, and `didOpen` is sent, then for some reason VSC thinks the connection is closed, so it restarts the server, but this time doesn't send `didOpen`. Something weird is going on. Maybe a bug in the stdio pipe or a bug in the Go LSP package.

**Comment by Timmmm** [comment]
Aha I figured it out! There was a bug in my `didOpen` handler that caused Go to panic, and for some reason I wasn't seeing the output (kind of hard to follow which pane all the debug messages go). That meant that VSC restarted the process, and the second time it *doesn't* send `didOpen`. Might be nice if it were documented somewhere how to get some kind of verbose logging from the client, so I could see a message like "stdin closed, restarting server"...

Anyway, I'll close this.

**Comment by dbaeumer** [comment]
@Timmmm there is a way to turn on tracing of the LSP client library. Use `${languageServerId}.trace.server`. For example to enable trace for the eslint server it is `"eslint.trace.server" = "messages"`.
